simplefi-stake-dao/src/stakeDAOVault.ts

```diff
@@ -0,0 +1,288 @@
+    deposit.lpTokenAmount = event.params.value
+    deposit.lpTokenTransferEventApplied = true
+    deposit.save()
+    handleDeposit(event, deposit)
+    return
+  }
+
+  // burn
+  if (toHex == ADDRESS_ZERO) {
+    let withdraw = getOrCreateWithdraw(event, vault, fromHex)
+    withdraw.lpTokenAmount = event.params.value
+    withdraw.lpTokenTransferEventApplied = true
+    withdraw.save()
+
+    vault.lastTransferToZero = event.transaction.hash.toHexString()
+    vault.save()
+
+    handleWithdraw(event, withdraw)
+    return
+  }
+
+  transferOutpuToken(
+    event,
+    vault,
+    event.params.from,
+    event.params.to,
+    event.params.value
+  )
+}
+
+export function handleInputTokenTransfer(event: InputTokenTransfer): void {
+  let vaultInputToken = VaultInputTokenEntity.load(event.address.toHexString()) as VaultInputTokenEntity
+  let vault = VaultEntity.load(vaultInputToken.vault) as VaultEntity
+  let fromHex = event.params.from.toHexString()
+  let toHex = event.params.to.toHexString()
+
+  // Deposit
+  if (toHex == vault.id) {
+    let deposit = getOrCreateDeposit(event, vault, fromHex)
+    deposit.inputTokenAmount = event.params.value
+    deposit.inputTokenTransferEventApplied = true
+    deposit.save()
+    handleDeposit(event, deposit)
+    return
+  }
+
+  // Withdraw
+  if (fromHex == vault.id && toHex != vault.controller) {
+    let withdraw = getOrCreateWithdraw(event, vault, toHex)
+    withdraw.inputTokenAmount = event.params.value
+    withdraw.inputTokenTransferEventApplied = true
+    withdraw.save()
+    handleWithdraw(event, withdraw)
+    return
+  }
+}
+
+function getOrCreateDeposit(event: ethereum.Event, vault: VaultEntity, account: string): DepositEntity {
+  let id = event.transaction.hash.toHexString()
+  let deposit = DepositEntity.load(id)
+  if (deposit != null) {
+    return deposit as DepositEntity
+  }
+
+  deposit = new DepositEntity(id)
+  deposit.vault = vault.id
+  deposit.account = account
+  deposit.inputTokenTransferEventApplied = false
+  deposit.lpTokenTransferEventApplied = false
+  deposit.save()
+  return deposit as DepositEntity
+}
+
+function getOrCreateWithdraw(event: ethereum.Event, vault: VaultEntity, account: string): WithdrawEntity {
+  let id = event.transaction.hash.toHexString()
+  let withdraw = WithdrawEntity.load(id)
+  if (withdraw != null) {
+    return withdraw as WithdrawEntity
+  }
+
+  withdraw = new WithdrawEntity(id)
+  withdraw.vault = vault.id
+  withdraw.account = account
+  withdraw.inputTokenTransferEventApplied = false
+  withdraw.lpTokenTransferEventApplied = false
+  withdraw.save()
+  return withdraw as WithdrawEntity
+}
+
+function handleDeposit(event: ethereum.Event, deposit: DepositEntity): void {
+  if (!deposit.inputTokenTransferEventApplied || !deposit.lpTokenTransferEventApplied) {
+    return
+  }
+
+  let vault = VaultEntity.load(deposit.vault) as VaultEntity
+  vault.totalSupply = vault.totalSupply.plus(deposit.lpTokenAmount as BigInt)
```

Comment:
Why is `totalSupply` updated based on the existing value in DB, but `balance` is updated by querying the contract directly? Just wondering in general what's the best practice?

Reply:
@gvladika Generally we prefer to use events and other triggers to keep track of contract state variables but in some cases where it's easy to use a contract call instead of a lot of events we use contract call. Reason behind avoiding contract calls is that it's order of magnitude expensive in terms of time consumed while running mapping code because it's an external API call. 

I tried to do it without contract call but it meant tracking controller, vault and strategy contract on lots of function call handlers and then finding correct logic to update balance value based on order of execution of these functions.

Reply:
Got it, thanks